flexmeasures/data/models/planning/storage.py

```diff
@@ -198,20 +199,66 @@ def _prepare(self, skip_validation: bool = False) -> tuple:  # noqa: C901
             soc_min,
         )
 
+        consumption_capacity = self.flex_model.get("consumption_capacity")
+        production_capacity = self.flex_model.get("production_capacity")
+
         if sensor.get_attribute("is_strictly_non_positive"):
             device_constraints[0]["derivative min"] = 0
         else:
-            device_constraints[0]["derivative min"] = power_capacity_in_mw * -1
+            device_constraints[0]["derivative min"] = (
+                -1
+            ) * get_continuous_series_sensor_or_quantity(
+                quantity_or_sensor=production_capacity,
+                actuator=sensor,
+                unit=sensor.unit,
+                query_window=(start, end),
+                resolution=resolution,
+                beliefs_before=belief_time,
+                fallback_attribute="production_capacity",
+                max_value=convert_units(power_capacity_in_mw, "MW", sensor.unit),
+            )
         if sensor.get_attribute("is_strictly_non_negative"):
             device_constraints[0]["derivative max"] = 0
         else:
-            device_constraints[0]["derivative max"] = power_capacity_in_mw
+            device_constraints[0][
+                "derivative max"
+            ] = get_continuous_series_sensor_or_quantity(
+                quantity_or_sensor=consumption_capacity,
+                actuator=sensor,
+                unit=sensor.unit,
+                query_window=(start, end),
+                resolution=resolution,
+                beliefs_before=belief_time,
+                fallback_attribute="consumption_capacity",
+                max_value=convert_units(power_capacity_in_mw, "MW", sensor.unit),
+            )
 
         # Apply round-trip efficiency evenly to charging and discharging
-        device_constraints[0]["derivative down efficiency"] = (
+        charging_efficiency = get_continuous_series_sensor_or_quantity(
+            quantity_or_sensor=self.flex_model.get("charging_efficiency"),
+            actuator=sensor,
+            unit="%",
+            query_window=(start, end),
+            resolution=resolution,
+            beliefs_before=belief_time,
+            fallback_attribute="charging-efficiency",
+        )
+        discharging_efficiency = get_continuous_series_sensor_or_quantity(
+            quantity_or_sensor=self.flex_model.get("discharging_efficiency"),
+            actuator=sensor,
+            unit="%",
+            query_window=(start, end),
+            resolution=resolution,
+            beliefs_before=belief_time,
+            fallback_attribute="discharging-efficiency",
+        )
+
+        device_constraints[0][
+            "derivative down efficiency"
+        ] = discharging_efficiency.fillna(roundtrip_efficiency**0.5)
+        device_constraints[0]["derivative up efficiency"] = charging_efficiency.fillna(
```

Comment:
I think these fields should represent mutually exclusive use cases. Either the user sets a round-trip efficiency, or they set the charging and discharging efficiencies. That would be a schema check.

Reply:
Yes, true, they are mutually exclusive.

I've added a check on the schema.

Reply:
This is not what I had in mind. I imagine there are cases where only a charging efficiency is relevant, and discharging is not possible at all (so it doesn't make sense to define its efficiency). What I was suggesting instead was to forbid these three cases:
1. flex-model contains both a round-trip efficiency and a charging efficiency
2. flex-model contains both a round-trip efficiency and a discharging efficiency
3. flex-model contains a round-trip efficiency, a charging efficiency and a discharging efficiency

Reply:
You're right, those the cases to be considered. I've added those in a new commit.